### PR TITLE
[expo-updates][iOS] Refactor check and fetch APIs, implement state machine

### PR DESCRIPTION
**Pull request by douglowder**
- Refactor checkForUpdateAsync and fetchUpdateAsync
- Fix inconsistencies in handling rollbacks
- New state machine implementation: `UpdatesStateMachine.swift`
- Fire native events into JS when state changes

To support the above, I added a few new delegate methods in `AppLoaderTaskDelegate` to allow us to be notified at different stages of checking for and downloading updates (including when assets are downloaded).

# Test Plan

- Existing E2E should pass
- New iOS unit tests for the state machine
- Add new event listener and hook to the Updates JS code to verify that state change events are received.

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-updates/ios/EXUpdates/AppController.swift [hunk 1]
```diff
@@ -38,12 +39,32 @@ public class AppController: NSObject, AppLoaderTaskDelegate, ErrorRecoveryDelega
   private static let ErrorDomain = "EXUpdatesAppController"
   private static let EXUpdatesEventName = "Expo.nativeUpdatesEvent"
 
+  // Events for the legacy UpdateEvent JS listener
   public static let UpdateAvailableEventName = "updateAvailable"
   public static let NoUpdateAvailableEventName = "noUpdateAvailable"
   public static let ErrorEventName = "error"
 
+  // Events that will be used to track state changes
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/AppController.swift:47]: What do you envision the state looking like? I had imagined it as being something like:
```
{  
  isCheckingForNewUpdate: bool;  
  newUpdate: Update | null; // next update that will be launched when restarted  
  newUpdateError: Error | null; // error that may have occurred  
  isNewUpdateDownloaded: bool; // whether the new update is downloaded, maybe this should be a property within newUpdate  
  newUpdateDownloadProgress: number | null; // maybe combine with `isNewUpdateDownloaded` and just have 100 mean downloaded
}
```

I think it may be worth using a state machine to keep track of the state. But that is the next PR I guess.

The reason I bring this up here in the review is that I think a lot of these events may break the abstraction direction a bit (AppController shouldn't know about the Module). I had imagined this more being purely a delegation pattern. Concretely, this means that all events would be emitted by the classes underneath this (by either delegate calls or nsnotificationcenter if necessary) and that this class would accumulate them. This may require some changes to the module though, so I'm not sure of the scope of this.

Just for example sake, here's how I imagined `fetchUpdateAsync` would work:
1. module calls `RemoteAppLoader.loadUpdate`
2. That remote app loader has a reference to AppController somehow (not sure if this is true? should we make it true if not? or does that break some abstraction in the interface? what even is the interface lol?)
3. The remote app loader does it's stuff and propagates events to its reference to AppController through its reference.

File: packages/expo-updates/ios/EXUpdates/ReactDelegateHandler/ExpoUpdatesAppDelegateSubscriber.swift [hunk 2]
```diff
@@ -15,4 +15,9 @@ public final class ExpoUpdatesAppDelegateSubscriber: ExpoAppDelegateSubscriber {
     }
     return true
   }
+
+  public func applicationWillEnterForeground(_ application: UIApplication) {
```
Comment (wschurman) [inline packages/expo-updates/ios/EXUpdates/ReactDelegateHandler/ExpoUpdatesAppDelegateSubscriber.swift:19]: this should be in a separate PR I think?

**Review by wschurman — COMMENTED** [review]
Will do a more thorough review after architecture questions are settled (I don't want to create unnecessary changes in this revision if the architecture is going to change).

File: packages/expo-updates/ios/EXUpdates/ReactDelegateHandler/ExpoUpdatesAppDelegateSubscriber.swift (hunk 2 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/ReactDelegateHandler/ExpoUpdatesAppDelegateSubscriber.swift:19]: Agreed, this was mainly to test if `checkForUpdate()` can be called from outside JS in response to a notification.

File: packages/expo-updates/ios/EXUpdates/AppController.swift (hunk 1 above)
Comment (douglowder) [reply packages/expo-updates/ios/EXUpdates/AppController.swift]: The state machine is now implemented in this PR.

File: packages/expo-updates/e2e/setup/project.js [hunk 3]
```diff
@@ -478,7 +478,7 @@ async function initAsync(
   // enable proguard on Android
   await fs.appendFile(
     path.join(projectRoot, 'android', 'gradle.properties'),
-    '\nandroid.enableProguardInReleaseBuilds=true',
+    '\nandroid.enableProguardInReleaseBuilds=true\nandroid.kotlinVersion=1.8.20',
```
Comment (brentvatne) [inline packages/expo-updates/e2e/setup/project.js:481]: given that you're using prebuild here, it would be better to use `expo-build-properties` to set the kotlin version via config plugin: https://docs.expo.dev/versions/latest/sdk/build-properties/#pluginconfigtypeandroid

Comment (douglowder) [reply packages/expo-updates/e2e/setup/project.js:481]: @brentvatne very sadly that does not work -- it seems that the Detox config plugin is getting applied _after_ the `expo-build-properties` plugin. This seems very strange, as my `plugins` array has `expo-build-properties` last.
